### PR TITLE
Collect metrics with respect to worker transitions

### PR DESCRIPTION
**Pull request by sundargates**
### Context

This diff aims to collect individual worker metrics such as 
1. How long it took to schedule a worker?
2. How long it took to prepare a worker?
3. How long a worker ran?

All these metrics are tagged with cluster names. This would help us differentiate mesos & titus schedule latencies, for instance. 

### Checklist

- [ ] `./gradlew build` compiles code correctly
- [ ] Added new tests where applicable
- [ ] `./gradlew test` passes all tests
- [ ] Extended README or added javadocs where applicable

### CODE REVIEW

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+    private final Duration epochDuration;
+    private final Clock clock;
+
+    @Override
+    protected void runOneIteration() throws Exception {
+        Instant expiry = clock.instant().minus(cleanupInterval);
+        Set<CleanupJobEvent> toBeRemoved = new HashSet<>();
+
+        synchronized (jobsToBeCleaned) {
```
Comment (hmitnflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:75]: wondering if you could use a ConcurrentHashMap for the set as well
and also using an iterator on jobToBeCleaned and call remove on the iterator to remove.w

**Review by hmitnflx — APPROVED** [review]

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java [hunk 2]
```diff
@@ -30,4 +31,9 @@ public void process(LifecycleEventsProto.WorkerListChangedEvent event) {
     public void process(LifecycleEventsProto.JobStatusEvent statusEvent) {
         logger.info("Received status event {}", statusEvent);
     }
+
+    @Override
+    public void process(WorkerStatusEvent workerStatusEvent) {
+        logger.info("Received worker status event {}", workerStatusEvent);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java:37]: Will this be quite chatty on info?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                .build();
+
+            m = MetricsRegistry.getInstance().registerAndGet(m);
+            this.schedulingDuration = m.getTimer(SCHEDULING_DURATION);
+            this.preparationDuration = m.getTimer(PREPARATION_DURATION);
+            this.runningDuration = m.getTimer(RUNNING_DURATION);
+        }
+
+        private void reportSchedulingDuration(long durationInMillis) {
+            this.schedulingDuration.record(durationInMillis, TimeUnit.MILLISECONDS);
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:191]: do we want to tag jobId etc.?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java (hunk 2 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java:37]: good point

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java (hunk 3 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:191]: I don't think atlas would be happy about that - atlas only supports a max cardinality of 1000. The goal is to look at metrics per resource cluster / mesos. For instance, we want to show that titus metrics are better or at least similar to mesos metrics.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java (hunk 1 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:75]: good point. let me do that.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java (hunk 2 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerEventSubscriberLoggingImpl.java]: done.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java (hunk 1 above)
Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java]: done.

**Review by hmitnflx — APPROVED** [review]

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+ */
+@RequiredArgsConstructor
+@Slf4j
+public class WorkerMetricsCollector extends AbstractScheduledService implements
+    WorkerEventSubscriber {
+
+    public static final String SCHEDULING_DURATION = "schedulingDuration";
+    public static final String PREPARATION_DURATION = "preparationDuration";
+    public static final String RUNNING_DURATION = "runningDuration";
+    private final ConcurrentMap<JobId, Map<WorkerId, IMantisWorkerMetadata>> jobWorkers =
```
Comment (calvin681) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:60]: How does this map get populated when Master first start or a new Master is elected?

Comment (sundargates) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:60]: The JobActor sends the WorkerListChangedEvent (https://github.com/Netflix/mantis/blob/d9cc7b67306b02d0e824cc6e5e0024f2475f7e21/mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/LifecycleEventsProto.java#L264) on initialization (https://github.com/Netflix/mantis/blob/d9cc7b67306b02d0e824cc6e5e0024f2475f7e21/mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java#L1460). That's how the map gets populated after a restart.

Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java:60]: Emit the size of this map.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java [hunk 5]
```diff
@@ -148,9 +152,17 @@ public MasterMain(ConfigurationFactory configFactory, AuditEventSubscriber audit
             final StatusEventSubscriber statusEventSubscriber = new StatusEventSubscriberAkkaImpl(statusEventBrokerActor);
             final AuditEventSubscriber auditEventSubscriberAkka = new AuditEventSubscriberAkkaImpl(auditEventBrokerActor);
             final WorkerEventSubscriber workerEventSubscriber = WorkerRegistryV2.INSTANCE;
+            final WorkerMetricsCollector workerMetricsCollector = new WorkerMetricsCollector(
+                Duration.ofMinutes(5), // cleanup jobs after 5 minutes
```
Comment (Andyz26) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/MasterMain.java:156]: nit configurable

**Review by Andyz26 — APPROVED** [review]
It's probably good to monitor the perf diff when releasing this one too.

**Pull request merged**